staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -130,11 +131,24 @@ import (
 // A delete notification exposes the last locally known non-absent
 // state, except that its ResourceVersion is replaced with a
 // ResourceVersion in which the object is actually absent.
+//
+// The informational methods (EventHandlerCount, IsStopped, IsStarted)
+// are intended to be used to manage informers in any upper informer
+// management layer for creating and destroying informers on-the fly when
+// adding or removing handlers.
+// Beware of race conditions: Such a layer must hide the basic informer
+// objects from its users and offer a closed *synchronized* view for informers.
+// Although these informational methods are synchronized each, in
+// sequences the state queried first might have been changed before
+// calling the next method, if other callers (or the stop channel)
+// are able to interact with the informer interface in parallel.
 type SharedInformer interface {
 	// AddEventHandler adds an event handler to the shared informer using the shared informer's resync
 	// period.  Events to a single handler are delivered sequentially, but there is no coordination
 	// between different handlers.
-	AddEventHandler(handler ResourceEventHandler)
+	// It returns a handle for the handler that can be used to remove
+	// the handler again.
+	AddEventHandler(handler ResourceEventHandler) (*ResourceEventHandlerHandle, error)
```

Comment:
I'm not sure it's legit to start returning an error; that will leave preexisting code likely still compiling but not handling the error.

Instead I think it is safer to add a new method ("AddRemovableEventHandler"). Same comment for below method.

All of these interface changes break implementors of the interface; we should either know that that is very rare, or we should make SharedInformerV2 interface.

Reply:
This method might have been failed even before (for stopped informers). But there was no way for a caller to react on this, there was just a log message.

With this change this behavior for actual callers in unchanged and the coding does not get worse. But now the caller has the change to figure out that something went wrong and what went wrong. He is still free to decide to ignore it. 

May be there are warnings in some environments now, that the error return is ignored, but this hint gives you the chance to remember that here the coding could be improved.

Because of the interface change I had started a discussion in the mailing list months ago, and it it was approved.

Reply:
I found that email thread, it looks like past!me was expecting an additional method in this same interface. But it looks like everyone was in favor so I won't insist.